rasterio/windows.py

```diff
@@ -88,5 +140,172 @@ def intersect(*windows):
+    x_max, y_max = transform * (col_max, row_min)
+    return x_min, y_min, x_max, y_max
+
+
+def crop_window(window, height, width):
+    """Returns a window cropped to fall within height and width."""
+    (r_start, r_stop), (c_start, c_stop) = window
+    return (
+        (min(max(r_start, 0), height), max(0, min(r_stop, height))),
+        (min(max(c_start, 0), width), max(0, min(c_stop, width))))
+
+
+def eval_window(window, height, width):
+    """Evaluates a window tuple that might contain negative values
+    in the context of a raster height and width."""
+    try:
+        r, c = window
+        assert len(r) == 2
+        assert len(c) == 2
+    except (ValueError, TypeError, AssertionError):
+        raise ValueError("invalid window structure; expecting ints"
+                         "((row_start, row_stop), (col_start, col_stop))")
+    r_start = r[0] or 0
+    if r_start < 0:
+        if height < 0:
+            raise ValueError("invalid height: %d" % height)
+        r_start += height
+    r_stop = r[1] or height
+    if r_stop < 0:
+        if height < 0:
+            raise ValueError("invalid height: %d" % height)
+        r_stop += height
+    if not r_stop >= r_start:
+        raise ValueError(
+            "invalid window: row range (%d, %d)" % (r_start, r_stop))
+    c_start = c[0] or 0
+    if c_start < 0:
+        if width < 0:
+            raise ValueError("invalid width: %d" % width)
+        c_start += width
+    c_stop = c[1] or width
+    if c_stop < 0:
+        if width < 0:
+            raise ValueError("invalid width: %d" % width)
+        c_stop += width
+    if not c_stop >= c_start:
+        raise ValueError(
+            "invalid window: col range (%d, %d)" % (c_start, c_stop))
+    return (r_start, r_stop), (c_start, c_stop)
+
+
+def get_index(x, y, affine, op=math.floor, precision=6):
+    """
+    Returns the (row, col) index of the pixel containing (x, y) given a
+    coordinate reference system.
+
+    Parameters
+    ----------
+    x : float
+        x value in coordinate reference system
+    y : float
+        y value in coordinate reference system
+    affine : tuple
+        Coefficients mapping pixel coordinates to coordinate reference system.
+    op : function
+        Function to convert fractional pixels to whole numbers (floor, ceiling,
+        round)
+    precision : int
+        Decimal places of precision in indexing, as in `round()`.
+
+    Returns
+    -------
+    row : int
+        row index
+    col : int
+        col index
+    """
+    # Use an epsilon, magnitude determined by the precision parameter
+    # and sign determined by the op function: positive for floor, negative
+    # for ceil.
+    eps = 10.0**-precision * (1.0 - 2.0*op(0.1))
+    row = int(op((y - eps - affine[5]) / affine[4]))
+    col = int(op((x + eps - affine[2]) / affine[0]))
+    return row, col
+
+
+def get_window(left, bottom, right, top, affine, precision=6):
```

Comment:
I think it's probably ok to drop `get_window` and inline it with `window` (which is the only function that calls it - no need for redundancy)